cmd/podman/common/completion.go

```diff
@@ -0,0 +1,638 @@
+		return nil, cobra.ShellCompDirectiveError
+	}
+
+	for _, container := range containers {
+		// include ids in suggestions if more then 2 chars are typed
+		if len(toComplete) > 1 {
+			suggestions = append(suggestions, container.ID[0:12])
+		}
+		// include name in suggestions
+		suggestions = append(suggestions, container.Names...)
+	}
+	return suggestions, cobra.ShellCompDirectiveNoFileComp
+}
+
+func getPods(status string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	suggestions := []string{}
+	listOpts := entities.PodPSOptions{
+		Filters: make(map[string][]string),
+	}
+
+	if status != "all" {
+		listOpts.Filters = map[string][]string{"status": {status}}
+	}
+
+	pods, err := registry.ContainerEngine().PodPs(registry.GetContext(), listOpts)
+	if err != nil {
+		cobra.CompErrorln(err.Error())
+		return nil, cobra.ShellCompDirectiveError
+	}
+
+	for _, pod := range pods {
+		// include ids in suggestions if more then 2 chars are typed
+		if len(toComplete) > 1 {
+			suggestions = append(suggestions, pod.Id[0:12])
+		}
+		// include name in suggestions
+		suggestions = append(suggestions, pod.Name)
+	}
+	return suggestions, cobra.ShellCompDirectiveNoFileComp
+}
+
+func getVolumes() ([]string, cobra.ShellCompDirective) {
+	suggestions := []string{}
+	lsOpts := entities.VolumeListOptions{}
+
+	volumes, err := registry.ContainerEngine().VolumeList(registry.GetContext(), lsOpts)
+	if err != nil {
+		cobra.CompErrorln(err.Error())
+		return nil, cobra.ShellCompDirectiveError
+	}
+
+	for _, volume := range volumes {
+		suggestions = append(suggestions, volume.Name)
+	}
+	return suggestions, cobra.ShellCompDirectiveNoFileComp
+}
+
+func getImages(toComplete string) ([]string, cobra.ShellCompDirective) {
+	suggestions := []string{}
+	listOptions := entities.ImageListOptions{}
+
+	images, err := registry.ImageEngine().List(registry.GetContext(), listOptions)
+	if err != nil {
+		cobra.CompErrorln(err.Error())
+		return nil, cobra.ShellCompDirectiveError
+	}
+
+	for _, image := range images {
+		// FIXME: need ux testing
+		// discuss when image ids should be completed
+		// include ids in suggestions if more then 2 chars are typed
+		if len(toComplete) > 1 {
+			suggestions = append(suggestions, image.ID[0:12])
+		}
+		for _, repo := range image.RepoTags {
+			if toComplete == "" {
+				// suggest only full repo path if no input is given
+				suggestions = append(suggestions, repo)
+			} else {
+				// suggested "registry.fedoraproject.org/f29/httpd:latest" as
+				// - "registry.fedoraproject.org/f29/httpd:latest"
+				// - "registry.fedoraproject.org/f29/httpd"
+				// - "f29/httpd:latest"
+				// - "f29/httpd"
+				// - "httpd:latest"
+				// - "httpd"
+				paths := strings.Split(repo, "/")
+				for i := range paths {
+					suggestionWithTag := strings.Join(paths[i:], "/")
+					suggestionWithoutTag := strings.SplitN(suggestionWithTag, ":", 2)[0]
+					suggestions = append(suggestions, suggestionWithTag, suggestionWithoutTag)
+				}
```

Comment:
This logic does not work correct for images pulled via digest. I will fix this.

I wonder what the best ux for image completion would be. Currently we only complete the full reference with and without tag. I personally would prefer if we also complete the short form, e.g. `alpine`. But this could result in many displayed suggestions. With the current shortnames work I think it would make sense to display all defined shortnames and all full references for the present images.

@vrothberg WDYT? Is it possible to retrieve all defined shortnames here?